### PR TITLE
Test integration of new device interface with finite shots and `qml.execute`

### PR DESCRIPTION
**Pull request by albi3ro**
Follow on from #3903 . Uses changes from #4114 

This PR just adds finite shots testing with parameter shift to the test matrix.

I also make the test parametrized over the device, so we can use the same file to test integration with lightning qubit as well.

### CODE REVIEW

**Comment by albi3ro** [comment]
[sc-38571]

**Review by timmysilv — APPROVED** [review]
looks great, straightforward enough!

File: tests/interfaces/default_qubit_2_integration/test_autograd_default_qubit_2.py [hunk 1]
```diff
@@ -155,99 +161,109 @@ def cost(a):
...
         assert isinstance(res, tuple) and len(res) == 2
         assert res[0].shape == (2,)
         assert res[1].shape == (2,)
 
         expected = ([-np.sin(a), np.sin(a) * np.sin(b)], [0, -np.cos(a) * np.cos(b)])
-        assert all(np.allclose(_r, _e, atol=tol, rtol=0) for _r, _e in zip(res, expected))
+        for _r, _e in zip(res, expected):
+            assert np.allclose(_r, _e, atol=atol_for_shots(shots))
 
     @pytest.mark.filterwarnings("ignore:Attempted to compute the gradient")
```
Comment (eddddddy) [inline tests/interfaces/default_qubit_2_integration/test_autograd_default_qubit_2.py:194]: Is this `filterwarnings` marker still necessary? Same as below.

**Review by eddddddy — APPROVED** [review]
💯

**Review by mudit2812 — APPROVED** [review]

File: tests/interfaces/default_qubit_2_integration/test_autograd_default_qubit_2.py (hunk 1 above)
Comment (albi3ro) [reply tests/interfaces/default_qubit_2_integration/test_autograd_default_qubit_2.py:194]: actually, not sure what this warning is about any more.

Comment (albi3ro) [reply tests/interfaces/default_qubit_2_integration/test_autograd_default_qubit_2.py:194]: Ah. The first tape has no trainable parameters.  So yes the warning is still there.

**Pull request merged**